src/win32/fs.cpp

```diff
@@ -1109,95 +1097,86 @@ size_t WinFileSystemAccess::lastpartlocal(const string* name) const
-
-    const wchar_t* ptr = (const wchar_t*)(filename->data() + filename->size() 
-        - (filename->size() & 1));   // if the string has had an extra null char added for surety, get back on wchar_t boundary.
+    const wchar_t* ptr = filenamePath.localpath.data() + filenamePath.localpath.size();
 
     char c;
     size_t i, j;
 
-	size--;
+    size--;
 
-	if (size * sizeof(wchar_t) > filename->size())
-	{
-		size = int(filename->size() / sizeof(wchar_t));
-	}
+    if (size > filenamePath.localpath.size())
+    {
+        size = int(filenamePath.localpath.size());
+    }
+
+    for (i = 0; i < size; i++)
+    {
+        if (*--ptr == '.')
+        {
+            for (j = 0; j <= i; j++)
+            {
+                if (*ptr < '.' || *ptr > 'z') return false;
 
-	for (i = 0; i < size; i++)
-	{
-		if (*--ptr == '.')
-		{
-			for (j = 0; j <= i; j++)
-			{
-				if (*ptr < '.' || *ptr > 'z') return false;
+                c = (char)*(ptr++);
 
-				c = (char)*(ptr++);
+                // tolower()
+                if (c >= 'A' && c <= 'Z') c |= ' ';
 
-				// tolower()
-				if (c >= 'A' && c <= 'Z') c |= ' ';
-                
                 extension[j] = c;
-			}
-			
+            }
+
             extension[j] = 0;
-            
-			return true;
-		}
-	}
 
-	return false;
+            return true;
+        }
+    }
+
+    return false;
 }
 
 bool WinFileSystemAccess::expanselocalpath(LocalPath& pathArg, LocalPath& absolutepathArg)
 {
-    string* path = pathArg.editStringDirect();
-    string* absolutepath = absolutepathArg.editStringDirect();
-
-    string localpath = *path;
-    localpath.append("", 1);
-
 #ifdef WINDOWS_PHONE
     wchar_t full[_MAX_PATH];
-    if (_wfullpath(full, (wchar_t *)localpath.data(), _MAX_PATH))
+    if (_wfullpath(full, pathArg.localpath.data(), _MAX_PATH))
     {
-        absolutepath->assign((char *)full, wcslen(full) * sizeof(wchar_t));
+        absolutepathArg.localpath.assign(full, wcslen(full));
         return true;
     }
-    *absolutepath = *path;
+    absolutepathArg.localpath = pathArg.localpath;
     return false;
 #else
 
-    int len = GetFullPathNameW((LPCWSTR)localpath.data(), 0, NULL, NULL);
+    int len = GetFullPathNameW(pathArg.localpath.data(), 0, NULL, NULL);
     if (len <= 0)
     {
-        *absolutepath = *path;
+        absolutepathArg.localpath = pathArg.localpath;
         return false;
     }
 
-    absolutepath->resize(len * sizeof(wchar_t));
-    int newlen = GetFullPathNameW((LPCWSTR)localpath.data(), len, (LPWSTR)absolutepath->data(), NULL);
+    absolutepathArg.localpath.resize(len);
+    int newlen = GetFullPathNameW(pathArg.localpath.data(), len, absolutepathArg.localpath.data(), NULL);
     if (newlen <= 0 || newlen >= len)
     {
-        *absolutepath = *path;
+        absolutepathArg.localpath = pathArg.localpath;
         return false;
     }
 
-    if (memcmp(absolutepath->data(), L"\\\\?\\", 8))
+    if (memcmp(absolutepathArg.localpath.data(), L"\\\\?\\", 8))
```

Comment:
This overuse of `memcmp` and hard coded byte count could really have been replaced, now that the code needed change anyway.

Reply:
replaced memcmp() with string::compare() at places especially ones with hard codes byte counts. There are still places where memcmp is getting used. I will get back to those later once the target changes are stable.

Reply:
I seem to remember there is something about string::compare, something about how it compares the end of the string - if one string finishes but the other doesn't, even if you specify the length to compare, then they won't be equal, I think, or something like that.  If you make this change that will need careful testing.  thanks

Reply:
Thanks for bringing it up. I will test that area and make necessary changes if required.